### PR TITLE
EPL-8486: Allow Consumers to Specify a Custom $ref Resolver

### PR DESCRIPTION
**Pull request by michaelcarter-wf**
## Ultimate problem:
- json_schema can only retrieve remote refs asynchronously from HTTP(S) URLs.

## How it was fixed:
- allow the consumer to pass in sync and async ref providers.

## Testing suggestions:
- ddev test
- CI

## Potential areas of regression:
- ref resolution

### CODE REVIEW

**Comment by semveraudit-wf** [comment]
### Public API Changes

 Recommendation: **:warning: Minor version bump**


```diff
@@ line 44: package:json_schema/src/json_schema/typedefs.dart @@
+  typedef Future<JsonSchema> RefProviderAsync(String ref);
// Adding a typedef is a minor change.
```

```diff
@@ line 43: package:json_schema/src/json_schema/typedefs.dart @@
+  typedef JsonSchema RefProvider(String ref);
// Adding a typedef is a minor change.
```

```diff
@@ line 65: package:json_schema/src/json_schema/json_schema.dart @@
class JsonSchema {}
-     JsonSchema createSchema(dynamic data, {String schemaVersion, Uri fetchedFromUri});
+     JsonSchema createSchema(dynamic data, {String schemaVersion, Uri fetchedFromUri, RefProvider refProvider});
//    `refProvider` was added.
//    Adding an optional parameter is a minor change.

-     Future<JsonSchema> createSchemaAsync(dynamic data, {String schemaVersion, Uri fetchedFromUri});
+     Future<JsonSchema> createSchemaAsync(dynamic data, {String schemaVersion, Uri fetchedFromUri, RefProviderAsync refProvider});
//    `refProvider` was added.
//    Adding an optional parameter is a minor change.

```

Showing results for 7065e4c6bd89b096a1cef0decff409d91fc1d305
> Powered by [semver-audit-service](https://github.com/Workiva/semver-audit-service). Please report any problems by [filing an issue](https://github.com/Workiva/semver-audit-service/issues).
> Reported by semver-audit-dart 1.4.0
[Browse public API](https://semver-audit.wk-dev.wdesk.org/Workiva/json_schema/7065e4c6bd89b096a1cef0decff409d91fc1d305).

**Comment by michaelcarter-wf** [comment]
need to add testing for async ref resolvers as well...

**Comment by michaelcarter-wf** [comment]
make sure schema version is always passed in correctly

**Comment by shawnrusaw-wf** [comment]
+1 code review

File: lib/src/json_schema/json_schema.dart [hunk 1]
```diff
@@ -1099,13 +1144,18 @@ class JsonSchema {
         return _addSchemaToRefMap(originalRef.toString(), schema);
       };
 
-      final AsyncRetrievalOperation refSchemaOperation =
-          () => createSchemaFromUrl(_ref.toString()).then(addSchemaFunction);
+      final AsyncRetrievalOperation asyncRefSchemaOperation = _refProviderAsync == null
```
[the remark this answers is not in the document]
Comment (michaelcarter-wf) [reply lib/src/json_schema/json_schema.dart]: done

**Comment by michaelcarter-wf** [comment]
+10 

- [x] CI passes

**Pull request merged**